README.md
# salt-nanny [![Build Status](https://travis-ci.org/dandb/salt-nanny.svg?branch=master)](https://travis-ci.org/dandb/salt-nanny) [![Coverage Status](https://coveralls.io/repos/github/dandb/salt-nanny/badge.svg?branch=master)](https://coveralls.io/github/dandb/salt-nanny?branch=master)
Python Module that parses salt returns stored in redis and logs output 

# Example Usage:
*Command Line Usage:*
```
salt-nanny localhost minion1 minion2
```
Look for salt returns in localhost for minions - minion1 & minion2

```
salt-nanny localhost minion1 minion2 -p 6380 -x 20 -I 5 60 2
```
This command tells salt-nanny to wait 5, 10, 20, 40 and 60 seconds between each retry initially and then 60s for
subsequent retries. Attempt 20 times and then give up. Use port 6380 for redis.

```
salt-nanny localhost minion1 -r -j 20170905144844350779
```
This command tells salt-nanny to return results from minion minion1 that have higher JID than 20170905144844350779

*Example Python code:*
```
#!/usr/bin/env python
import salt.client
from saltnanny import SaltNanny

# Initialize SaltNanny with the cache & salt function
config = {'type': 'redis', 'host':'localhost', 'port':6379, 'db':'0'}
nanny = SaltNanny(config, 'test', 'state.highstate')

# Use SaltNanny to track returns to the external job cache
salt_nanny.initialize(['minion1', 'minion2'])
salt_nanny.track_returns()
return_code = salt_nanny.process_returns()
```

* For the example above, the log file in logs/test-state.highstate.log will contain results of the salt highstate
* The return code is 0 if all the salt functions for all minions succeded with a return dict containing retcode:0.
* SaltNanny also checks state results in case of a highstate. If any one state fails, the retcode is non zero.


CHANGES
Release Notes
=============
v0.1.1 First working version\n
v0.1.2 Make changes in redis key structure for salt 2016 compatibility\n
v0.1.3 Fix bug when reading the redis return\n
v0.1.4 Add facility to read the last redis return without tracking a currently running highstate.\n
v0.1.5 Fix formatting of log message.\n
v0.1.6 Add salt-nanny command line utility to track returns from running highstate.\n
v0.1.7 Update Documentation & improve command line utility\n
v0.1.9 Fixed bug to return exit code 2 if no returns are found.
v0.2.0 Added new command line parameter(earliest_jid) to only return saltmaster results more recent than the passed jid
v0.2.1 Updated list of potential_failures in salt_return_parser
v0.3.0 python 3 compatibility




tests/saltnanny_test.py
import os
import unittest
from mock import MagicMock
from mock import call
from mock import patch
from fakeredis import FakeRedis

from saltnanny import SaltNanny


class SaltNannyTest(unittest.TestCase):

    cache_config = {
        'type': 'redis',
        'host': 'localhost',
        'port': 6379,
        'db': '0'
    }

    @patch('redis.Redis')
    def test_initialize(self, mock_redis):
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client = MagicMock()
        salt_nanny.initialize(['minion1', 'minion2'])
        salt_nanny.cache_client.get_latest_jid.assert_has_calls([call('minion1', 'state.highstate'), call('minion2', 'state.highstate')])

    @patch('redis.Redis')
    def test_initialize_nominions(self, mock_redis):
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client = MagicMock()
        self.assertRaises(ValueError, salt_nanny.initialize, [])

    @patch('redis.Redis')
    def test_track_returns(self, mock_redis):
        fake_redis = FakeRedis()

        # Create and initialize Salt Nanny
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client.redis_instance = fake_redis
        salt_nanny.initialize(['minion1', 'minion2'])
        salt_nanny.min_interval = 1

        # Make Redis Returns available in fake redis
        fake_redis.set('minion1:state.highstate', '1234')
        fake_redis.set('minion2:state.highstate', '4321')
        fake_redis.hset('ret:1234', 'minion1', 'Highstate Result1')
        fake_redis.hset('ret:1234', 'minion2', 'Highstate Result1')

        # Start tracking returns
        salt_nanny.track_returns()

    @patch('redis.Redis')
    def test_parse_last_return(self, mock_redis):
        fake_redis = FakeRedis()

        # Create and initialize Salt Nanny
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client.redis_instance = fake_redis
        salt_nanny.initialize(['minion1'])
        salt_nanny.min_interval = 1

        with open('{0}/resources/highstate.json'.format(os.path.dirname(__file__)), 'r') as f:
            json = f.read()

        # Make Redis Returns available in fake redis
        fake_redis.set('minion1:state.highstate', '6789')
        fake_redis.set('minion1:6789', json)

        self.assertTrue(salt_nanny.parse_last_return() > 0)

    @patch('redis.Redis')
    def test_parse_last_return_with_results(self, mock_redis):
        fake_redis = FakeRedis()

        # Create and initialize Salt Nanny
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client.redis_instance = fake_redis
        salt_nanny.initialize(['minion1'])
        salt_nanny.min_interval = 1

        with open('{0}/resources/highstate.json'.format(os.path.dirname(__file__)), 'r') as f:
            json = f.read()

        # Make Redis Returns available in fake redis
        fake_redis.set('minion1:state.highstate', '6789')
        fake_redis.set('minion1:6789', json)

        # 6789 is higher than JID 6666 so the json will be fetched
        self.assertTrue(salt_nanny.parse_last_return(6666) == 2)

    @patch('redis.Redis')
    def test_parse_last_return_with_no_results(self, mock_redis):
        fake_redis = FakeRedis()

        # Create and initialize Salt Nanny
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client.redis_instance = fake_redis
        salt_nanny.initialize(['minion1'])
        salt_nanny.min_interval = 1

        with open('{0}/resources/highstate.json'.format(os.path.dirname(__file__)), 'r') as f:
            json = f.read()

        # Make Redis Returns available in fake redis
        fake_redis.set('minion1:state.highstate', '6789')
        fake_redis.set('minion1:6789', json)

        # 6789 is lower than JID 7777 so no results expected
        self.assertTrue(salt_nanny.parse_last_return(7777) == 1)

    @patch('redis.Redis')
    def test_track_custom_event_failures_no_failures(self, mock_redis):
        fake_redis = FakeRedis()
        fake_redis.set('custom_event_type', '["Random Event Log", "Success"]')
        # Create and initialize Salt Nanny
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client.redis_instance = fake_redis
        salt_nanny.min_interval = 1
        return_code = salt_nanny.track_custom_event_failures('custom_event_type', ['Failure'], 2)
        self.assertTrue(return_code == 0)

    @patch('redis.Redis')
    def test_track_custom_event_failures(self, mock_redis):
        fake_redis = FakeRedis()
        fake_redis.set('custom_event_type', '["Random Event Log", "Failure"]')
        # Create and initialize Salt Nanny
        salt_nanny = SaltNanny(self.cache_config)
        salt_nanny.cache_client.redis_instance = fake_redis
        salt_nanny.min_interval = 1
        return_code = salt_nanny.track_custom_event_failures('custom_event_type', ['Failure'], 2)
        self.assertTrue(return_code > 0)

    @patch('redis.Redis')
    def test_get_wait_time(self, mock_redis):
        salt_nanny = SaltNanny(self.cache_config)
        self.assertEquals(15, salt_nanny.get_wait_time(0))
        self.assertEquals(30, salt_nanny.get_wait_time(1))
        self.assertEquals(60, salt_nanny.get_wait_time(2))
        self.assertEquals(60, salt_nanny.get_wait_time(3))

    @patch('redis.Redis')
    def test_setup_logging(self, mock_redis):
        salt_nanny = SaltNanny(self.cache_config, 'test')
        self.assertTrue(salt_nanny.log is not None)


tests/test_saltnanny.py
import os
import pytest
import mock
from mock import MagicMock
from mock import call
from fakeredis import FakeRedis

from saltnanny import SaltNanny


@pytest.fixture(autouse=True)
def cache_config():
    yield {
        'type': 'redis',
        'host': 'localhost',
        'port': 6379,
        'db': '0'
    }


@pytest.fixture(autouse=True)
def salt_nanny(cache_config):
    with mock.patch('redis.Redis'):
        salt_nanny = SaltNanny(cache_config)

    salt_nanny.cache_client.redis_instance = FakeRedis()
    salt_nanny.min_interval = 1

    yield salt_nanny


def test_initialize(salt_nanny):
    salt_nanny.cache_client = MagicMock()
    salt_nanny.initialize(['minion1', 'minion2'])
    salt_nanny.cache_client.get_latest_jid.assert_has_calls([call('minion1', 'state.highstate'), call('minion2', 'state.highstate')])


def test_initialize_nominions(salt_nanny):
    salt_nanny.cache_client = MagicMock()

    with pytest.raises(ValueError):
        salt_nanny.initialize([])


# def test_track_returns_minion_2(salt_nanny):
#     fake_redis = salt_nanny.cache_client.redis_instance
#     fake_redis.set('minion1:state.highstate', '1234')
#
#     salt_nanny.initialize(['minion1'])
#
#     # Make Redis Returns available in fake redis
#     fake_redis.set('minion1:state.highstate', '4321')
#     fake_redis.hset('ret:4321', 'minion1', 'Highstate Result1')
#     fake_redis.hset('ret:1234', 'minion2', 'Highstate Result1')
#
#     # Start tracking returns
#     assert salt_nanny.track_returns(2) == 0


def test_track_returns_no_pending_minion_list(salt_nanny):
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('minion1:state.highstate', '1234')
    fake_redis.set('minion2:state.highstate', '1234')

    salt_nanny.initialize(['minion1', 'minion2'])
    salt_nanny.completed_minions = {'minion1': 'completed 1'}

    # Make Redis Returns available in fake redis
    fake_redis.set('minion2:state.highstate', '4321')
    fake_redis.hset('ret:1234', 'minion1', 'Highstate Result1')
    fake_redis.hset('ret:4321', 'minion2', 'Highstate Result1')

    # Start tracking returns
    assert salt_nanny.track_returns(2) == 0


def test_track_returns_no_match(salt_nanny):
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('minion1:state.highstate', '1234')

    salt_nanny.initialize(['minion1', 'minion2'])

    # Make Redis Returns available in fake redis
    fake_redis.set('minion2:state.highstate', '4321')
    fake_redis.hset('ret:1234', 'minion1', 'Highstate Result1')
    fake_redis.hset('ret:1234', 'minion2', 'Highstate Result1')

    # Start tracking returns
    assert 2 == salt_nanny.track_returns(2)


def test_track_returns(salt_nanny):
    salt_nanny.initialize(['minion1', 'minion2'])

    # Make Redis Returns available in fake redis
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('minion1:state.highstate', '1234')
    fake_redis.set('minion2:state.highstate', '4321')
    fake_redis.hset('ret:1234', 'minion1', 'Highstate Result1')
    fake_redis.hset('ret:1234', 'minion2', 'Highstate Result1')

    # Start tracking returns
    salt_nanny.track_returns(1)


def test_parse_last_return(salt_nanny):
    salt_nanny.initialize(['minion1'])

    with open('{0}/resources/highstate.json'.format(os.path.dirname(__file__)), 'r') as f:
        json = f.read()

    # Make Redis Returns available in fake redis
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('minion1:state.highstate', '6789')
    fake_redis.set('minion1:6789', json)

    assert salt_nanny.parse_last_return() > 0


def test_parse_last_return_with_results(salt_nanny):
    salt_nanny.initialize(['minion1'])

    with open('{0}/resources/highstate.json'.format(os.path.dirname(__file__)), 'r') as f:
        json = f.read()

    # Make Redis Returns available in fake redis
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('minion1:state.highstate', '6789')
    fake_redis.set('minion1:6789', json)

    # 6789 is higher than JID 6666 so the json will be fetched
    assert salt_nanny.parse_last_return(6666) == 2


def test_parse_last_return_with_no_results(salt_nanny):
    salt_nanny.initialize(['minion1'])

    with open('{0}/resources/highstate.json'.format(os.path.dirname(__file__)), 'r') as f:
        json = f.read()

    # Make Redis Returns available in fake redis
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('minion1:state.highstate', '6789')
    fake_redis.set('minion1:6789', json)

    # 6789 is lower than JID 7777 so no results expected
    assert salt_nanny.parse_last_return(7777) == 1


def test_track_custom_event_failures_no_failures(salt_nanny):
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('custom_event_type', '["Random Event Log", "Success"]')
    return_code = salt_nanny.track_custom_event_failures('custom_event_type', ['Failure'], 2)
    assert return_code == 0


def test_track_custom_event_failures(salt_nanny):
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('custom_event_type', '["Random Event Log", "Failure"]')
    return_code = salt_nanny.track_custom_event_failures('custom_event_type', ['Failure'], 2)

    assert return_code > 0


def test_track_custom_event_failures_key_not_found(salt_nanny):
    fake_redis = salt_nanny.cache_client.redis_instance
    fake_redis.set('custom_event_type', '["Random Event Log", "Failure"]')
    return_code = salt_nanny.track_custom_event_failures('event_no_match', ['Failure'], 1)

    assert return_code == 1


def test_get_wait_time(cache_config):
    with mock.patch('redis.Redis'):
        salt_nanny = SaltNanny(cache_config, 'test')
    assert salt_nanny.get_wait_time(0) == 15
    assert salt_nanny.get_wait_time(1) == 30
    assert salt_nanny.get_wait_time(2) == 60
    assert salt_nanny.get_wait_time(3) == 60


def test_setup_logging(cache_config):
    with mock.patch('redis.Redis'):
        salt_nanny = SaltNanny(cache_config, 'test')

    assert salt_nanny.log is not None

@mock.patch('os.makedirs')
@mock.patch('os.path.exists')
def test_setup_logging_make_logfile(mock_exists, mock_makedirs, cache_config):
    mock_exists.return_value = False
    mock_makedirs.return_value = False
    with mock.patch('redis.Redis'):
        salt_nanny = SaltNanny(cache_config, 'missing_log_dir')

    assert salt_nanny.log is not None
    mock_exists.assert_called_once_with(os.path.join(os.getcwd(), 'logs'))


tests/salt_nanny_tool_test.py
import unittest
from copy import copy
from argparse import ArgumentParser
from saltnanny import salt_nanny_tool, SaltNanny
from mock import patch, call


class ScriptsTest(unittest.TestCase):

    class MockArguments(object):
        pass

    mock_args = MockArguments()
    mock_args.host = 'localhost'
    mock_args.port = 6379
    mock_args.type = 'redis'
    mock_args.intervals = [1, 30, 2]
    mock_args.log_file = None
    mock_args.custom_event = 'state.highstate'
    mock_args.minions = ['minion1']
    mock_args.max_attempts = 15
    mock_args.last_return = False
    mock_args.earliest_jid = 20170905144844350780

    mock_args_return = copy(mock_args)
    mock_args_return.last_return = True

    @patch.object(ArgumentParser, 'add_argument')
    @patch.object(ArgumentParser, 'parse_args', return_value=mock_args)
    @patch.object(SaltNanny, '__init__', return_value=None)
    @patch.object(SaltNanny, 'initialize')
    @patch.object(SaltNanny, 'track_returns')
    def test_tool_main(self, mock_track_returns, mock_initialize, mock_init, mock_parse_args, mock_add_argument):
        salt_nanny_tool.tool_main()
        assert(mock_add_argument.call_count == 11)
        assert(mock_parse_args.call_count == 1)
        mock_track_returns.assert_called_once()

    @patch.object(ArgumentParser, 'add_argument')
    @patch.object(ArgumentParser, 'parse_args', return_value=mock_args_return)
    @patch.object(SaltNanny, '__init__', return_value=None)
    @patch.object(SaltNanny, 'initialize')
    @patch.object(SaltNanny, 'parse_last_return')
    def test_tool_main_return(self, mock_parse_last_return, mock_init, mock_initialize, mock_parse_args, mock_add_argument):
        salt_nanny_tool.tool_main()
        assert(mock_add_argument.call_count == 11)
        assert(mock_parse_args.call_count == 1)
        mock_parse_last_return.assert_called_once()


saltnanny/salt_nanny.py
import logging
import math
import os
from time import sleep


from .salt_return_parser import SaltReturnParser
from .salt_nanny_client import SaltNannyClientFactory


class SaltNanny:

    log = logging.getLogger('saltnanny')
    log_fmt = '%(asctime)s %(levelname)s %(message)s'

    def __init__(self, cache_config, target_log_file=None, fun='state.highstate',
                 min_interval=15, max_interval=60, multiplier=2):
        self.fun = fun
        self.initial_jids = {}
        self.completed_minions = {}
        self.min_interval = min_interval
        self.max_interval = max_interval
        self.multiplier = multiplier
        self.setup_logging(target_log_file)
        self.cache_client = SaltNannyClientFactory.factory(
            cache_config['type'], cache_config['host'], cache_config['port'], cache_config['db'])
        self.parser = SaltReturnParser(self.cache_client)

    def initialize(self, minion_list):
        if minion_list:
            self.minion_list = minion_list
            for minion in minion_list:
                self.initial_jids[minion] = self.cache_client.get_latest_jid(minion, self.fun)
        else:
            raise ValueError("No Minions provided. SaltNanny has no minions to babysit.")

    def track_returns(self, max_attempts=17):
        pending_minion_list = list(self.minion_list)

        for i in range(0, max_attempts):
            pending_minion_list = [minion for minion in pending_minion_list if minion not in list(self.completed_minions.keys())]

            if pending_minion_list:
                wait_time = self.get_wait_time(i)
                self.log.info('Sleeping for {0} seconds...'.format(wait_time))
                sleep(wait_time)
                self.log.info('SaltNanny is checking for minion returns in External Job Cache - Attempt: {0}'.format(i))
                for minion in pending_minion_list:
                    latest_jid = self.cache_client.get_latest_jid(minion, self.fun)
                    if latest_jid == self.initial_jids[minion]:
                        self.log.info('SaltNanny did not find a new JID for minion: {0}'.format(minion))
                        break
                    else:
                        self.log.info('SaltNanny found a New JID for minion: {0} JID: {1}'.format(minion, latest_jid))
                        self.completed_minions[minion] = latest_jid
            else:
                self.log.info('Results available in External Job Cache for all minions: {0}'
                              .format(list(self.completed_minions.keys())))
                break
        self.log.info(self.completed_minions)
        return self.parser.process_jids(self.completed_minions, len(self.minion_list))

    def parse_last_return(self, earliest_jid=0):
        for minion in self.minion_list:
            latest_jid = self.cache_client.get_latest_jid(minion, self.fun)
            if int(latest_jid) < earliest_jid:
                self.log.info("No highstates found in job cache for minion: {0} after: {1}".format(minion, earliest_jid))
                return 1
            else:
                self.completed_minions[minion] = latest_jid
        self.log.info(self.completed_minions)
        return self.parser.process_jids(self.completed_minions, len(self.minion_list))

    def get_wait_time(self, index):
        wait_interval = self.min_interval * math.pow(self.multiplier, index)
        if wait_interval > self.max_interval:
            return self.max_interval
        return wait_interval

    def track_custom_event_failures(self, event_key, failures, max_attempts=17, successes=None):
        if successes is None:
            successes = []
        for i in range(0, max_attempts):
            wait_time = self.get_wait_time(i)
            self.log.info('Sleeping for {0} seconds...'.format(wait_time))
            sleep(wait_time)
            self.log.info('SaltNanny is checking for Custom Event: {0} in External Job Cache - Attempt: {1}'
                          .format(event_key, i))
            if self.cache_client.exists(event_key):
                self.log.info('SaltNanny found the Custom Event: {0} in External Job Cache. Checking for failures...'
                              .format(event_key))
                return self.parser.check_custom_event_failure(event_key, failures, successes)
        self.log.info('SaltNanny has timed out waiting for Custom Event: {0}'.format(event_key))
        return 1

    def setup_logging(self, target_log_file):
        formatter = logging.Formatter(self.log_fmt)
        console_handler = logging.StreamHandler()
        console_handler.setFormatter(formatter)
        self.log.addHandler(console_handler)
        self.log.setLevel(logging.INFO)
        if target_log_file:
            target = os.path.join(os.getcwd(), 'logs/{0}-{1}.log'.format(target_log_file, self.fun))
            if not os.path.exists(os.path.dirname(target)):
                os.makedirs(os.path.dirname(target))
            ch = logging.FileHandler(target)
            ch.setFormatter(formatter)
            self.log.addHandler(ch)


saltnanny/salt_nanny_client.py
import abc
import logging
import redis


class SaltNannyClientFactory(object):

    log = logging.getLogger('saltnanny')

    @staticmethod
    def factory(cache_client_type, host, port, db):
        if cache_client_type == 'redis':
            return SaltRedisClient(host, port, db)
        else:
            SaltNannyClientFactory.log.error('Unrecognized Cache Client type: {0}'.format(cache_client_type))


class SaltNannyClient:
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def get_latest_jid(self, minion, fun):
        """ expected to return the most current jid for the specified minion, fun combination """

    @abc.abstractmethod
    def get_return_by_jid(self, minion, jid):
        """ expected to return the salt returner results for the specified minion,jid combination"""

    @abc.abstractmethod
    def get_value_by_key(self, base_key):
        """ expected to return the salt returner results associated with the key """

    @abc.abstractmethod
    def exists(self, base_key):
        """ expected to return true if the key exists, false otherwise """


class SaltRedisClient(SaltNannyClient):

    log = logging.getLogger('saltnanny')

    def __init__(self, host='localhost', port=6379, db='0'):
        try:
            self.redis_instance = redis.Redis(host, port, db)
            self.redis_instance.ping()
        except Exception as e:
            self.log.error('Error while trying to connect to redis cache: {0}'.format(e))
            raise e

    def _decode(self, value):
        if value is None:
            return value
        return value.decode('utf-8')

    def get_latest_jid(self, minion, fun):
        cache_key = '{0}:{1}'.format(minion, fun)
        latest_jid = '0'
        try:
            key_type = self._decode(self.redis_instance.type(cache_key))
            if key_type == 'list':
                latest_jid = self._decode(self.redis_instance.lindex(cache_key, 0))
            elif key_type == 'string':
                latest_jid = self._decode(self.redis_instance.get(cache_key))
        except KeyError:
            self.log.info('Latest jid not found. Defaulting to 0')
            pass
        self.log.info('Latest jid for Minion:{0} JID:{1}'.format(minion, latest_jid))
        return latest_jid

    def get_return_by_jid(self, minion, jid):
        cache_key = '{0}:{1}'.format(minion, jid)
        cache_key_salt_2016 = 'ret:{0}'.format(jid)
        if self.redis_instance.exists(cache_key):
            return self._decode(self.redis_instance.get(cache_key))
        elif self.redis_instance.exists(cache_key_salt_2016):
            return self._decode(self.redis_instance.hget(cache_key_salt_2016, minion))
        else:
            msg = 'Return info for JID:{0} does not exist'.format(jid)
            self.log.error(msg)
            raise ValueError(msg)

    def get_value_by_key(self, base_key):
        return self._decode(self.redis_instance.get(base_key))

    def exists(self, base_key):
        return self.redis_instance.exists(base_key)


saltnanny/salt_return_parser.py
import re
import logging
import json
from time import sleep
from ast import literal_eval


class SaltReturnParser:

    log = logging.getLogger('saltnanny')
    fun_running_pattern = 'is running as PID'

    def __init__(self, cache_client, min_interval=15, max_attempts=15):
        self.cache_client = cache_client
        self.min_interval = min_interval
        self.max_attempts = max_attempts

    def process_jids(self, completed_minions, all_minions_count):
        return_code_sum = 0
        for minion, jid in list(completed_minions.items()):
            try:
                return_info, return_code = self.get_return_info(minion, jid)
                return_code_sum += return_code
                self.log.info(json.dumps(return_info, indent=1))
            except ValueError as ve:  # pragma: no cover
                self.log.error('Error retrieving results for Minion:{0}: Exception:{1}'.format(minion, ve))

        if not completed_minions:
            self.log.info('No highstates found in Job Cache.')
            return 2

        if len(completed_minions) != all_minions_count and return_code_sum == 0:
            self.log.info('Highstates available in Job Cache were successful, timed out waiting for others.')
            return_code_sum = 1
        elif return_code_sum != 0:
            self.log.info('One or more highstates were not entirely successful. Please investigate.')
        else:
            self.log.info('All Highstates completed successfully!')

        return return_code_sum

    def check_custom_event_failure(self, cache_key, failures, successes):
        custom_results = literal_eval(self.cache_client.get_value_by_key(cache_key))
        self.log.info('Custom Event Return in Job Cache. Key: {0} Value:'.format(cache_key))
        if isinstance(custom_results, list):
            # Print results on each line if its a list (for example a list of log statements)
            for result in custom_results:
                self.log.info(result)

            for result in custom_results:
                if self.check_successes(result, successes):
                    return 0
                if self.check_failures(result, failures):
                    return 1
        else:
            self.log.info(custom_results)
            if self.check_successes(custom_results, successes):
                return 0
            if self.check_failures(custom_results, failures):
                return 1
        return 0

    @staticmethod
    def check_failures(result, failures):
        failures_exist = [True for failure in failures if failure in result]
        return True in failures_exist

    @staticmethod
    def check_successes(result, successes):
        successes_exist = [True for success in successes if success in result]
        return True in successes_exist

    def get_return_info(self, minion, jid, attempt=1):
        self.log.info('Getting return info for Minion:{0} JID:{1}'.format(minion, jid))

        return_info = self.cache_client.get_return_by_jid(minion, jid)
        return_dict = json.loads(return_info)
        return_code = return_dict.get('retcode')

        if self.is_fun_running(return_dict) and attempt < self.max_attempts:
            self.log.info('Return Info for JID:{0} indicates that the function is still running.'.format(jid))
            self.log.info('Sleeping for {0} seconds...'
                          .format(self.min_interval, jid))
            sleep(self.min_interval)
            attempt += 1
            return self.get_return_info(minion, jid, attempt)

        if self.highstate_failed(return_info) or not isinstance(return_code, int):
            return_code = 1

        return return_dict, return_code

    def highstate_failed(self, result):
        try:
            possible_failures = [
                '"result": false',
                'Data failed to compile:',
                'Pillar failed to render with the following messages:',
                'Detected conflicting IDs',
                'Cannot extend ID'
            ]
            failures = [failure in result for failure in possible_failures]
            self.log.info(failures)
            if True not in failures:
                failures = self.check_regex_failure(failures, result)
            return True in failures
        except:  # pragma: no cover
            self.log.error('Error finding if there was a failure in the result:\n {0}'.format(result))
            return True

    def check_regex_failure(self, failures, result):
        regex_failure = r"Rendering SLS '.*' failed:"
        failures.append(bool(re.search(regex_failure, result)))
        return failures

    def is_fun_running(self, return_dict):
        if 'return' in return_dict and isinstance(return_dict['return'], list):
            for return_item in return_dict['return']:
                if self.fun_running_pattern in return_item:
                    return True
        return False


saltnanny/salt_nanny_tool.py
from argparse import ArgumentParser
from .salt_nanny import SaltNanny


def get_args():
    parser = ArgumentParser()
    parser.add_argument('host', type=str, help='Redis Host that stores salt returns.')
    parser.add_argument('minions', type=str, nargs='+', help='List of Highstated Minions')
    parser.add_argument('-t', '--type', action='store', dest='type', type=str, default='redis', help='Type of external job cache(redis/etc).')
    parser.add_argument('-p', '--port', action='store', dest='port', type=int, default=6379, help='Redis port.')
    parser.add_argument('-l', '--log', action='store', dest='log_file', type=str, default=None, help='Log file.')
    parser.add_argument('-e', '--custom-event', action='store', dest='custom_event', type=str, default='state.highstate', help='Custom Event if this is not a highstate')
    parser.add_argument('-x', '--max-attempts', action='store', dest='max_attempts', type=int, default=15, help='Custom Event if this is not a highstate')
    parser.add_argument('-I', '--intervals', action='store', dest='intervals', type=int, nargs=3, default=[15, 60, 2], help='Custom Wait intervals and multiplier')
    parser.add_argument('-r', '--last-return', action='store_true', dest='last_return', help='Fetch last highstate result.')
    parser.add_argument('-j', '--earliest-jid', action='store', dest='earliest_jid', nargs='?', default=0, type=int, help='Fetch last highstate result stored after this time')
    return parser.parse_args()


def tool_main():
    args = get_args()
    cache_config = {'type': args.type, 'host': args.host, 'port': args.port, 'db': '0'}
    salt_nanny = SaltNanny(cache_config, args.log_file, args.custom_event, args.intervals[0], args.intervals[1], args.intervals[2])
    salt_nanny.initialize(args.minions)
    return salt_nanny.parse_last_return(args.earliest_jid) if args.last_return else salt_nanny.track_returns(args.max_attempts)


def main():  # pragma: no cover
    exit(tool_main())


if __name__ == '__main__':  # pragma: no cover
    main()
